### PR TITLE
SSHArc Public Preview improvements

### PR DESCRIPTION
**Pull request by vthiebaut10**
- Improve error message for proxy errors:
image
- Improve help
- Don't look for SSH under System32. Assume Path is set correctly.
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your PR is merged into master branch, a new PR will be created to update `src/index.json` automatically.  
The precondition is to put your code inside this repo and upgrade the version in the PR but do not modify `src/index.json`.

### CODE REVIEW

File: src/ssh/azext_ssh/_help.py [hunk 1]
```diff
@@ -13,13 +13,13 @@
 helps['ssh vm'] = """
     type: command
     short-summary: SSH into Azure VMs or Arc Servers.
-    long-summary: Users can login using AAD issued certificates or using local user credentials. We recommend login using AAD issued certificates. To SSH as a local user in the target machine, you must provide the local user name using the --local-user argument.
+    long-summary: Users can login using AAD issued certificates or using local user credentials. We recommend login using AAD issued certificates. To SSH using local user credentials, you must provide the local user name using the --local-user parameter.
     examples:
-        - name: Give a resource group and VM to SSH using AAD issued certificates
+        - name: Give a resource group name and resource name to SSH using AAD issued certificates
```
Comment (bagajjal) [inline src/ssh/azext_ssh/_help.py:18]: `resource name -> machine name` to make it more readable

File: src/ssh/azext_ssh/_help.py [hunk 2]
```diff
@@ -32,36 +32,66 @@
...
-            az ssh vm --local-user username --resource-group myResourceGroup --vm-name myVM --private-key-file key
+            az ssh vm --local-user username --resource-group myResourceGroup --name myVM --private-key-file key
 
-        - name: Give a local user name to SSH using local user credentials on the target machine using password based authentication.
+        - name: Give a local user name to SSH with local user credentials using password based authentication.
           text: |
-            az ssh vm --local-user username --resource-group myResourceGroup --vm-name myArcServer
+            az ssh vm --local-user username --resource-group myResourceGroup --name myArcServer
+        
+        - name: Give a SSH Client Folder to use the ssh executables (ssh-keyge.exe, ssh.exe, etc) in that folder. If not provided, the extension will look for pre-installed ssh client (for Windows, it will look for ssh executables under C:\\Windows\\System32\\OpenSSH).
```
Comment (bagajjal) [inline src/ssh/azext_ssh/_help.py:54]: Give a SSH Client Folder to use SSH executables like ssh-keyge.exe, ssh.exe, etc.

File: src/ssh/azext_ssh/_help.py [hunk 3]
```diff
@@ -32,36 +32,66 @@
...
+          text: |
+            az ssh vm --resource-group myResourceGroup --name myVM --ssh-client-folder "C:\\Program Files\\OpenSSH"
+        
+        - name: When connecting to an Arc Server, optionally give a SSH Proxy Folder to indicate the folder where the SSH proxy will be stored. If not provided, the proxy will be saved in .clientsshproxy folder in user\'s home directory.
```
Comment (bagajjal) [inline src/ssh/azext_ssh/_help.py:58]: let's delete this one. There are very few customers who are interested in this option. They can look at the possible arguments field.

File: src/ssh/azext_ssh/_help.py [hunk 4]
```diff
@@ -32,36 +32,66 @@
...
+          text: |
+            az ssh config --resource-group myResourceGroup --name myVm --file ./sshconfig
+            ssh -F ./sshconfig myResourceGroup-myVM
+
+        - name: Give a resource group, resource name, and local user to create a config using local user credentials, save in local file, and then ssh into that resource
           text: |
-            az ssh config --resource-group myResourceGroup --vm-name myVm --file ./sshconfig
+            # password based authentication
+            az ssh config --resource-group myResourceGroup --name myVM --local-user username1 --file ./sshconfig
+            ssh -F ./sshconfig MyResourceGroup-myVM-username1
```
Comment (bagajjal) [inline src/ssh/azext_ssh/_help.py:77]: delete line 77, 81. Keep line 85.
Make sure the username is same at line 76,80,83

File: src/ssh/azext_ssh/_help.py [hunk 5]
```diff
@@ -32,36 +32,66 @@
...
+            # certificate based authentication
+            az ssh config --resource-group myResourceGroup --name myVM --local-user username3 --certificate-file cert.pub --private-key-file key --file ./sshconfig
+            ssh -F ./sshconfig MyResourceGroup-myVM-username3
 
         - name: Give the public IP (or hostname) of a VM for which to create a config and then ssh
           text: |
             az ssh config --ip 1.2.3.4 --file ./sshconfig
             ssh -F ./sshconfig 1.2.3.4
+        
+        - name: Give Keys Destination Folder to indicate where new generated keys and certificates will be saved. If not provided, keys will be saved to a new folder "az_ssh_config" next to the config file.
```
Comment (bagajjal) [inline src/ssh/azext_ssh/_help.py:92]: Give Keys Destination Folder to store the generated keys and certificates. If not provides, SSH keys are stored in new folder "az_ssh_config" next to the config file.

File: src/ssh/azext_ssh/ssh_utils.py [hunk 6]
```diff
@@ -143,6 +145,9 @@ def get_ssh_cert_principals(cert_file, ssh_client_folder=None):
 
 
 def _print_error_messages_from_ssh_log(log_file, connection_status, delete_cert):
+    if connection_status:
+        connection_status = connection_status.returncode
```
Comment (bagajjal) [inline src/ssh/azext_ssh/ssh_utils.py:149]: This is dangerous. You are changing the type of connection_status from object to int.
Use a different variable names.

File: src/ssh/azext_ssh/ssh_utils.py [hunk 7]
```diff
@@ -302,27 +307,37 @@ def _start_cleanup(cert_file, private_key_file, public_key_file, delete_credenti
...
-            temp_dir = os.path.dirname(cert_file)
-            file_utils.delete_folder(temp_dir, f"Couldn't delete temporary folder {temp_dir}", True)
+    try:
+        if connection_status: 
+            if connection_status.returncode != 0:
+                # Check if stderr is a proxy error
+                regex = ("{\"level\":\"fatal\",\"msg\":\"sshproxy: error copying information from the connection: "
+                         ".*\",\"time\":\".*\"}.*")
+                if re.search(regex, connection_status.stderr):
+                    logger.error("SSH Proxy Error. Check if incoming connections are enabled on the hybrid agent "
```
Comment (bagajjal) [inline src/ssh/azext_ssh/ssh_utils.py:317]: logger.error("Please make sure SSH port is allowed and SSH server is running on the target machine")

File: src/ssh/azext_ssh/ssh_utils.py [hunk 8]
```diff
@@ -363,6 +378,8 @@ def _issue_config_cleanup_warning(delete_cert, delete_keys, is_arc, cert_file, r
 
 
 def _get_connection_status(log_file, connection_status):
+    if connection_status:
+        connection_status = connection_status.returncode
```
Comment (bagajjal) [inline src/ssh/azext_ssh/ssh_utils.py:382]: This is dangerous. You are changing the type of connection_status from object to int.
Use a different variable names.

**Review by bagajjal — CHANGES_REQUESTED** [review]
.

File: src/ssh/azext_ssh/_help.py (hunk 4 above)
Comment (vthiebaut10) [reply src/ssh/azext_ssh/_help.py:77]: I was trying to show in that example that you can create config entries for different local users in the target machine, and be able to differentiate between them with the username appended to the Host name.

Comment (vthiebaut10) [reply src/ssh/azext_ssh/_help.py:74]: I changed it to include one example with rg and name, and one with ip

File: src/ssh/azext_ssh/ssh_utils.py [hunk 9]
```diff
@@ -55,11 +55,23 @@ def start_ssh_connection(op_info, delete_keys, delete_cert):
...
+            if set(['-v', '-vv', '-vvv']).isdisjoint(ssh_arg_list):
+                connection_status = subprocess.run(command, env=env, stderr=subprocess.PIPE, text=True)
+            else:
+                # Logs are sent to stderr. In that case, we shouldn't capture stderr.
+                connection_status = subprocess.run(command, env=env, text=True)
+        except Exception as e:
+            recommendation = ("Ensure OpenSSH is installed, and that the PATH Environment Variable is set correctly.\n"
+                              "Alternatively, use --ssh-client-folder to provide folder path with OpenSSH executables.")
+            raise azclierror.BadRequestError(f"Failed to run ssh command with error: {str(e)}.",
+                                             recommendation)
```
Comment (vthiebaut10) [inline src/ssh/azext_ssh/ssh_utils.py:70]: @bagajjal What do you think of this error message?

**Pull request merged**